policy: SelectorCache integration to policy computation

jrajahalme opened the pull request:
```release-note
Policy label matching is now cached instead of repeated from scratch on each identity add/delete.
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7958)
<!-- Reviewable:end -->

ianvernon commented on pkg/envoy/server.go:
```diff
@@ -589,15 +589,14 @@ func createBootstrap(filePath string, name, cluster, version string, xdsSock, eg
-			_, ok := deniedIdentities[id]
-			if !ok && sel.Matches(labels) {
+			if sel.Matches(labels) {
+				log.Infof("Selector %v matched labels %v (for id %d)", sel, labels, uint64(id))
```
This should be debug level, or removed.

ianvernon commented on pkg/envoy/server.go:
```diff
@@ -697,6 +697,8 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, policyEnforced bool,
 					break
 				}
+				i++
+				log.Infof("NP GOt rule (%d), sel: %v, l7: %v, rule: %v", i, sel, l7, rule)
```
Clean up this log message (set to debug level, clean up message formatting), or remove it.

ianvernon commented on pkg/kvstore/allocator/allocator.go:
```diff
@@ -566,6 +566,7 @@ func (a *Allocator) Allocate(ctx context.Context, key AllocatorKey) (idpool.ID,
 		value, isNew, err = a.lockedAllocate(ctx, key)
 		if err == nil {
 			a.mainCache.insert(key, value)
+			log.WithField(fieldKey, key).WithField(fieldID, value).Info("Allocated key")
```
Info --> Debug

ianvernon commented on pkg/policy/repository_test.go:
```diff
@@ -1844,7 +1844,7 @@ Ingress verdict: allowed
 	expectedOut = `
 Resolving ingress policy for [any:foo]
-0/1 rules selected
+0/0 rules selected
```
Shouldn't the last number always be the # of rules in the repository?

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -87,6 +89,33 @@ type identitySelector interface {
 	notifyUsers(added, deleted []identity.NumericIdentity)
 }
 
+// Identity is the information we need about a an identity that rules can select
```
about a an --> about an

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -118,13 +147,34 @@ var (
+	if len(l.namespaces) > 0 {
+		if ns != "" {
+			for i := range l.namespaces {
+				if ns == l.namespaces[i] {
```
This check is only for equality. K8s also has the notion of set-based membership - https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-based-requirement . Given that a policy always applies in a given namespace, I'm assuming we can only rely upon pure equality of strings instead of set-based membership operations. It'd be good to add a comment indicating as such.

ianvernon commented on daemon/policy_test.go:
```diff
@@ -272,10 +272,10 @@ func (ds *DaemonSuite) TestUpdateConsumerMap(c *C) {
-						RemotePolicies: expectedRemotePolicies,
-						L7:             &PNPAllowGETbar,
-					},
+					//{
```
Why is this commented out now?

jrajahalme replied:
I changed the l4 rule processing to always include any selectors in the l7rulesPerEp also in CachedSelectors, and never include duplicates in CachedSelectors, and figured that this would be OK as it is shadowed by the rule above.
I think this change is due to not having duplicate selectors in the former `Endpoints`. Like 85% sure :-)

ianvernon replied:
Yeah, as long as we are redirecting to the proxy, we should be OK, since the shadowing is done at L7 itself.

ianvernon commented on pkg/endpoint/policy.go:
```diff
@@ -120,32 +118,25 @@ func (e *Endpoint) regeneratePolicy(owner Owner) (retErr error) {
 	stats.waitingForPolicyRepository.End(true)
 
+	// Collect label arrays before policy computation, as this can fail.
+	// GH-1128 should allow optimizing this away, but currently we can't
```
This comment is really old - is it still relevant?

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -87,6 +89,33 @@ type identitySelector interface {
 }
 
+// Identity is the information we need about a an identity that rules can select
+type Identity struct {
```
Does this need to be exported?

jrajahalme replied:
renamed as `scIdentity`

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -87,6 +89,33 @@ type identitySelector interface {
+}
+
+// IdentityCache is a cache of identity to labels mapping
+type IdentityCache map[identity.NumericIdentity]Identity
```
Does this need to be exported?

jrajahalme replied:
Renamed as `scIdentityCache`

ianvernon commented on pkg/policy/rule_test.go:
```diff
@@ -1961,15 +1934,6 @@ func (ds *PolicyTestSuite) TestEgressL3AllowAllEntity(c *C) {
-
-	// Check that no CIDR rules were generated
-	c.Assert(len(cidrPolicy.Ingress.Map), Equals, 0)
-	c.Assert(len(cidrPolicy.Egress.Map), Equals, 0)
```
This commit could also be broken out into its own PR.

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -199,7 +234,28 @@ type fqdnSelector struct {
+	namespaces []string // allowed namespaces, or nil
+}
+
+func (l *labelIdentitySelector) MatchesNamespace(ns string) bool {
```
I'd rather this be in its own PR as well, as it's orthogonal to the actual integration of the usage of SelectorCache into the policy subsystem.

ianvernon commented on pkg/identity/cache/allocator.go:
```diff
@@ -174,7 +174,15 @@ func IdentityAllocationIsLocal(lbls labels.Labels) bool {
 // an identity for the specified set of labels already exist, the identity is
 // re-used and reference counting is performed, otherwise a new identity is
 // allocated via the kvstore.
-func AllocateIdentity(ctx context.Context, lbls labels.Labels) (*identity.Identity, bool, error) {
```
Add documentation to this function about the semantics surrounding usage of the owner; when is it acceptable to provide a nil owner vs. not?

ianvernon replied:
@jrajahalme ^

ianvernon commented on pkg/identity/cache/allocator.go:
```diff
@@ -202,32 +210,39 @@ func AllocateIdentity(ctx context.Context, lbls labels.Labels) (*identity.Identi
 // identity again. This function may result in kvstore operations.
 // After the last user has released the ID, the returned lastUse value is true.
-func Release(ctx context.Context, id *identity.Identity) (bool, error) {
+func Release(owner IdentityAllocatorOwner, ctx context.Context, id *identity.Identity) (bool, error) {
```
Add documentation to this function about the semantics surrounding usage of the owner; when is it acceptable to provide a nil owner vs. not?

ianvernon commented on pkg/identity/cache/cache.go:
```diff
@@ -188,6 +188,7 @@ func (w *identityWatcher) watch(owner IdentityAllocatorOwner, events allocator.A
 				}
 			}
 			// Issue collected updates
+			log.Infof("UPDATE IDENTITIES(added: %v, deleted: %v)", added, deleted)
```
Make this debug level, and use `WithFields`.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -130,16 +130,17 @@ func (l4 *L4Filter) HasL3DependentL7Rules() bool {
-		return !ok
-	default:
-		return true
+		// loop to get access to the first and only key in the map
```
To confirm - this is relying on the semantics that if the wildcard selector is in `L7RulesPerEp`, it will be the only key in `L7RulesPerEp` ?

joestringer replied:
I'm also curious, what's the practical difference between this and the previous version? (Is there a unit test case that can highlight this?)

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -163,8 +164,9 @@ func (l4 *L4Filter) ToKeys(direction trafficdirection.TrafficDirection, identity
-		identities := getSecurityIdentities(identityCache, &sel)
+	for _, cs := range l4.CachedSelectors {
+		identities := cs.GetSelections()
+		log.Debugf("ToKeys: Allowed remote IDs for selector %v: %v", cs, identities)
```
Use `WithFields` here.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -181,36 +183,39 @@ func (l4 *L4Filter) ToKeys(direction trafficdirection.TrafficDirection, identity
-		}
+// IdentitySelectionUpdated implements CachedSelectionUser interface
+func (l4 *L4Filter) IdentitySelectionUpdated(selector CachedSelector, selections, added, deleted []identity.NumericIdentity) {
+	log.Infof("L4Filter::IdentitySelectionUpdated(selector: %v, selections: %v, added: %v, deleted: %v) call received",
```
Info --> Debug?

Also recommend using `WithFields` here.

ianvernon replied:
Also, this function doesn't do anything because we are creating `L4Filters` and are not doing incremental updates - would be good to note that here, as it's not clear why we aren't doing anything.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -220,41 +225,47 @@ func (l7 L7DataMap) addRulesForEndpoints(rules api.L7Rules, endpoints []api.Endp
 // rules via the `rule` parameter.
-func CreateL4Filter(peerEndpoints api.EndpointSelectorSlice, rule api.PortRule, port api.PortProtocol,
-	protocol api.L4Proto, ruleLabels labels.LabelArray, ingress bool) L4Filter {
+func createL4Filter(peerEndpoints api.EndpointSelectorSlice, rule api.PortRule, port api.PortProtocol,
```
This is outside the scope of this PR, but we should strongly consider changing the parameters of these functions to be some type of utility struct containing all the parameters wrapped in a single type, because it keeps growing.

joestringer replied:
This isn't entirely trivial, because this function is the convergence point for creation of both ingress and egress filters. If the code immediately prior to creating this function has to assemble all of the fields into a struct, is it really any better than having a half dozen parameters? The point of this function is to create a structure with all of the specified parameters, in the form of the `L4Filter`.

(Though `protocol` here seems entirely redundant to me, `port` already contains a protocol.)

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -220,41 +225,47 @@ func (l7 L7DataMap) addRulesForEndpoints(rules api.L7Rules, endpoints []api.Endp
 	} else {
-		// If there are no explicit fromEps, have a 'special' wildcard endpoint.
-		l7[api.WildcardEndpointSelector] = rules
+		panic("Must have at least the wildcard selector")
```
Oof, panic? Why not return an error and fail accordingly in callers?

joestringer replied:
If it's practical to throw all paths of `L7DataMap` construction in some tests then I'm not super worried about this, but I agree that seeing the `panic` here makes not only the code panic, but also the developers ;)

ianvernon commented on pkg/policy/l4_test.go:
```diff
@@ -177,7 +180,7 @@ func (s *PolicyTestSuite) TestJSONMarshal(c *C) {
   "l7-rules": [
     {
-      "any.foo=": {
+      "\u0026LabelSelector{MatchLabels:map[string]string{any.foo: ,},MatchExpressions:[],}": {
```
This format is not very user friendly :(

ianvernon commented on pkg/policy/rule.go:
```diff
@@ -58,19 +58,44 @@ func (r *rule) String() string {
-		existingFilter.Endpoints = api.EndpointSelectorSlice{api.WildcardEndpointSelector}
+	if existingFilter.AllowsAllAtL3() {
+		// Release references held by filterToMerge.CachedSelectors
+		selectorCache.RemoveSelectors(filterToMerge, filterToMerge.CachedSelectors)
```
I think that this "remove and set to nil" functionality should be put in some utility function since it's crucial we have the two collocated together.

ianvernon commented on pkg/policy/rule.go:
```diff
@@ -58,19 +58,44 @@ func (r *rule) String() string {
 }
 
-func mergePortProto(ctx *SearchContext, endpoints []api.EndpointSelector, existingFilter, filterToMerge *L4Filter) error {
+func mergePortProto(ctx *SearchContext, existingFilter, filterToMerge *L4Filter, selectorCache *SelectorCache) error {
```
Even though this function is not exported, I think it'd be beneficial to add unit tests for this function due to the high importance of how we manage objects in the `SelectorCache` here.

ianvernon commented on pkg/proxy/kafka.go:
```diff
@@ -204,41 +203,31 @@ func createKafkaRedirect(r *Redirect, conf kafkaConfiguration, endpointInfoRegis
-		return false
-	} else {
-		flowdebug.Log(scopedLog.WithField("rule", string(b)), "Applying rule")
+	if flowdebug.Enabled() {
```
Add a comment that we are checking this explicitly to avoid Marshaling `rules` for no reason.

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -37,11 +38,52 @@ type CachedSelector interface {
+	return false
+}
+
+// Insert in a sorted order? Returns true if inserted, false if cs was already in
```
Are you asking whether we should insert in a given order or not?

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -153,6 +194,12 @@ type selectorManager struct {
 	cachedSelections map[identity.NumericIdentity]struct{}
 }
 
+// Equal is used by checker.Equals, and only considers the identity of the selector,
```
identity --> key?

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -238,7 +302,13 @@ type labelIdentitySelector struct {
-func (l *labelIdentitySelector) MatchesNamespace(ns string) bool {
+// xxxMatches returns true if the CachedSelector matches given labels.
+// This is slow, but only used for policy tracing, so it's OK.
+func (l *labelIdentitySelector) xxxMatches(labels labels.LabelArray) bool {
```
Why `xxx`?

jrajahalme replied:
To remind that this is not a cached match. Maybe "slowMatches" would be better?

ianvernon commented on pkg/policy/selectorcache.go:
```diff
@@ -477,50 +583,84 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache) {
 	}
 }
 
+// XXXGetIDCacheRevision can be used to figure our if the selections
```
Why XXX?

our --> out

jrajahalme replied:
This should not be necessary when we move to incremental updates. "XXX" to remind that this is supposed to be a temporary solution.

ianvernon reviewed (commented):
Solid work! I think that each commit of this PR would benefit from being its own PR, but that's not a strict requirement. If CI breakage does occur and is hard to track, it would be good to split out the commits separately for a smaller debugging surface.

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -87,6 +89,33 @@ type identitySelector interface {
+type Identity struct {
+	NID       identity.NumericIdentity
+	lbls      labels.LabelArray
+	namespace string // value of the namespace label, or nil
```
How can a string be "nil"? :)

jrajahalme replied:
Good catch :-) replacing `nil` with `""`.

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -124,6 +153,12 @@ type selectorManager struct {
 }
 
+//
+// CachedSelector implementation (== Public API)
```
What's this comment referring to?

jrajahalme replied:
Code following the comment :-)

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -199,7 +234,28 @@ type fqdnSelector struct {
 	selectorManager
-	selector api.EndpointSelector
+	selector   api.EndpointSelector
+	namespaces []string // allowed namespaces, or nil
```
Is there any reason why this isn't a map? The `MatchesNamespace` wouldn't need to iterate over the slice of namespaces.

jrajahalme replied:
My understanding is that in wast majority of the cases this will contain exactly one string. If so, a map would be slower.

aanm commented on pkg/policy/repository.go:
```diff
@@ -167,6 +167,8 @@ func (p *Repository) ResolveL4IngressPolicy(ctx *SearchContext) (*L4PolicyMap, e
 // rule found in the repository takes precedence.
+//
+// NOTE: This is only called from unit tests.
 func (p *Repository) ResolveL4EgressPolicy(ctx *SearchContext) (*L4PolicyMap, error) {
```
Then let's move this to the unit tests where the function would receive 2 arguments, a `*Repository` and a `*SearchContext`

jrajahalme replied:
Apparently there is no need to change the function definition at all, it worked just by cut&paste to `repository_test.go`.

aanm commented on pkg/policy/repository.go:
```diff
@@ -219,6 +213,8 @@ func (p *Repository) AllowsIngressRLocked(ctx *SearchContext) api.Decision {
 // held.
+//
+// NOTE: This is only called from unit tests.
 func (p *Repository) AllowsEgressRLocked(ctx *SearchContext) api.Decision {
```
Same here

jrajahalme replied:
This is used from unit tests of multiple packages, updated the comment to state that.

aanm commented on pkg/identity/cache/allocator.go:
```diff
@@ -174,7 +174,15 @@ func IdentityAllocationIsLocal(lbls labels.Labels) bool {
 // re-used and reference counting is performed, otherwise a new identity is
 // allocated via the kvstore.
-func AllocateIdentity(ctx context.Context, lbls labels.Labels) (*identity.Identity, bool, error) {
+func AllocateIdentity(owner IdentityAllocatorOwner, ctx context.Context, lbls labels.Labels) (id *identity.Identity, allocated bool, err error) {
```
good to know, can you address this @jrajahalme ?

ianvernon replied:
Addressed in follow-up commit.

aanm commented on pkg/identity/cache/allocator.go:
```diff
@@ -202,32 +210,39 @@ func AllocateIdentity(ctx context.Context, lbls labels.Labels) (*identity.Identi
 // identity again. This function may result in kvstore operations.
 // After the last user has released the ID, the returned lastUse value is true.
-func Release(ctx context.Context, id *identity.Identity) (bool, error) {
+func Release(owner IdentityAllocatorOwner, ctx context.Context, id *identity.Identity) (bool, error) {
```
Same here

ianvernon replied:
Addressed in follow-up commit.

aanm commented on pkg/identity/cache/allocator.go:
```diff
@@ -238,20 +253,26 @@ func Release(ctx context.Context, id *identity.Identity) (bool, error) {
 // where several identities may be allocated and another error means that they
 // should all be released.
-func ReleaseSlice(ctx context.Context, identities []*identity.Identity) error {
+func ReleaseSlice(owner IdentityAllocatorOwner, ctx context.Context, identities []*identity.Identity) error {
```
Same here

ianvernon replied:
Addressed in follow-up commit.

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -400,6 +470,16 @@ func (sc *SelectorCache) AddFQDNSelector(user CachedSelectionUser, fqdnSelec api
+func (sc *SelectorCache) FindCachedIdentitySelector(selector api.EndpointSelector) CachedSelector {
+	key := selector.LabelSelector.String()
+	sc.mutex.Lock()
+	defer sc.mutex.Unlock()
```
drop the defer, wrap line 479 between the locks

jrajahalme replied:
OK.

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -469,6 +546,35 @@ func (sc *SelectorCache) RemoveSelector(user CachedSelectionUser, selector Cache
+// updates on the cached selector.
+func (sc *SelectorCache) ChangeUser(from, to CachedSelectionUser, selector CachedSelector) {
+	sc.mutex.Lock()
+	defer sc.mutex.Unlock()
```
as far as I can tell the locks only need to be between line 571

jrajahalme replied:
The following lines manipulating the `users` maps also need the locks. However, I removed the defer and moved the unlock to the end of the function.

aanm commented on pkg/policy/selectorcache.go:
```diff
@@ -477,50 +583,84 @@ func (sc *SelectorCache) UpdateIdentities(added, deleted cache.IdentityCache) {
-			log.WithFields(logrus.Fields{logfields.Identity: numericID}).Warning("UpdateIdentities: Adding an existing identity")
+		if old, exists := sc.idCache[numericID]; exists {
+			// Skip if no change
+			if lbls.String() == old.lbls.String() {
```
I would strongly suggest compare this using an hash function... or something else that does not require building strings

jrajahalme replied:
Good point, added `LabelArray.Same()` that compares the labels in order, which should be good enough as the labels are sorted for the kv-store anyway.

joestringer commented on pkg/policy/resolve.go:
```diff
@@ -140,24 +147,23 @@ func NewEndpointPolicy() *EndpointPolicy {
 
 // Realizes copies the fields from desired into p. It assumes that the fields in
-// desired are not modified after this function is called.
+// desired are not modified or otherwise used after this function is called.
```
What does this mean, you can't read from the resulting 'desired' fields at all?

joestringer replied:
Ah, looks like this is deleted in a followup commit.

joestringer commented on pkg/endpoint/endpoint.go:
```diff
@@ -1083,8 +1082,8 @@ func ParseEndpoint(strEp string) (*Endpoint, error) {
-	ep.desiredPolicy = policy.NewEndpointPolicy()
-	ep.realizedPolicy = policy.NewEndpointPolicy()
+	ep.desiredPolicy = policy.NewEndpointPolicy(repo)
+	ep.realizedPolicy = policy.NewEndpointPolicy(repo)
```
Why do these need to be non-`nil` to begin with? Policy resolution will overwrite them, and realizing these shouldn't be triggered until after the first policy resolution, no?

jrajahalme replied:
Long story short (I tried to make them nil before), they contain members that *may* be set before they are first set. Could revisit this later when have more time, though.

ianvernon replied:
I'm okay with doing this to get this merged in, but am a bit weary of having to make the endpoint aware of the repository. These things should ideally be separate ? We should reconsider in future work what we really want the policy structures within `Endpoint` to be. I think that since the creator / manager of policies is going to be external to the endpoint (e.g., whatever generates `SelectorPolicy`) that we should not have to make the Endpoint aware of the repository. This is out of scope of this PR, though.

joestringer commented on pkg/ipcache/cidr.go:
```diff
@@ -58,10 +58,11 @@ func allocateCIDRs(prefixes []*net.IPNet) ([]*identity.Identity, error) {
 		}
-
-		id, isNew, err := cache.AllocateIdentity(context.Background(), cidr.GetCIDRLabels(prefix))
+		// XXX: Figure out if CIDR identity allocations need to be propagated to the selector cache.
```
If CIDR identity allocations aren't propagated into the selector cache, then presumably we can't fully switch to using the selectorcache to derive all identity selections by rules?

jrajahalme replied:
I think the events are still propagated to the selector cache via the kvstore watcher events.

jrajahalme replied:
Or maybe not?
Looks like pkg/ipcache/kvstore.go is only used for unit testing?

joestringer commented on pkg/policy/l4.go:
```diff
@@ -220,41 +225,47 @@ func (l7 L7DataMap) addRulesForEndpoints(rules api.L7Rules, endpoints []api.Endp
 		Ingress:          ingress,
 	}
 
+	if len(peerEndpoints) == 0 || peerEndpoints.SelectsAllEndpoints() {
```
> `len(peerEndpoints) == 0`

This interprets the `peerEndpoints` in the L4 way (allow all), not the L3 way (deny all). Is there earlier code ensuring that L3 deny-all rules don't get to this point? I thought that the L4 path already injects a wildcard selector at the appropriate point higher in the policy generation logic, which renders this check unnecessary.

Put another way, the previous version of this function handled this logic differently, what is the functional change of the `len(peerEndpoints)` check added here?

ianvernon replied:
Would be keen on a comment here indicating why this is necessary as well.

jrajahalme replied:
Did not know this is never called with an empty `peerEndpoints`, removed the length == 0 check.

jrajahalme replied:
Unit tests violate this rule, will remove the calls with an empty `peerEndpoints`.

joestringer commented on pkg/policy/repository.go:
```diff
@@ -124,22 +133,23 @@ func wildcardL3L4Rule(proto api.L4Proto, port int, endpoints api.EndpointSelecto
-		filter.Endpoints = append(filter.Endpoints, endpoints...)
 		filter.DerivedFromRules = append(filter.DerivedFromRules, ruleLabels)
-		l4Policy[k] = filter
+		// l4Policy[k] = filter // pointer now, no need to reset
```
Do we need to retain this line?

jrajahalme replied:
No :-)

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -469,6 +546,35 @@ func (sc *SelectorCache) RemoveSelector(user CachedSelectionUser, selector Cache
+
+// ChangeUser changes the CachedSelectionUser that gets updates on the
+// updates on the cached selector.
+func (sc *SelectorCache) ChangeUser(from, to CachedSelectionUser, selector CachedSelector) {
```
nit: Usually, functions which manipulate or act upon some parameter pass that parameter as the first parameter. I was surprised by the ordering here, so ended up having to look up the implementation multiple times as in the callers it's not so simple as `ChangeUser(from, to, selector)` (as they don't name the from/to as from/to). Similarly with `RemoveSelector()` and friends, the key is usually the first parameter.

jrajahalme replied:
Flipped these.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -58,19 +58,44 @@ func (r *rule) String() string {
+		selectorCache.RemoveSelectors(existingFilter, existingFilter.CachedSelectors)
+		existingFilter.CachedSelectors = nil
 		existingFilter.allowsAllAtL3 = true
+		for _, cs := range filterToMerge.CachedSelectors {
```
This hunk and the next hunk look identical, can they be shared?

I don't think that the existing `Case 1`, `Case 2` comments in this function make sense any more as the implementation significantly changes here.

jrajahalme replied:
Done sharing, updated comments to match the code.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -58,19 +58,44 @@ func (r *rule) String() string {
+		for _, cs := range filterToMerge.CachedSelectors {
+			if existingFilter.CachedSelectors.Insert(cs) {
+				// Update selector owner to the existingFilter
+				selectorCache.ChangeUser(filterToMerge, existingFilter, cs)
```
Did you consider adding an extra pass through the filters after all merging to handle `selectorCache` references, rather than adding them upon creation and passing the selector cache through all of the merge logic?

jrajahalme replied:
No. We are now actually using the cached selectors in some parts of the logic. E.g., checking if peer endpoints select the Host, for example.

joestringer commented on pkg/policy/rule_test.go:
```diff
@@ -419,29 +420,32 @@ func (ds *PolicyTestSuite) TestMergeL4PolicyEgress(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(res, Not(IsNil))
-	c.Assert(*res, checker.DeepEquals, *expected)
+	c.Assert(*res, checker.Equals, *expected)
```
I think that all of the `DeepEquals` -> `Equals` changes could be split out to reduce the size of this patch.

jrajahalme replied:
Will do if more rounds are needed.

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -37,11 +38,52 @@ type CachedSelector interface {
 }
 
+// CachedSelectorSlice is a slice of CachedSelectors that can be sorted.
+type CachedSelectorSlice []CachedSelector
```
This type introduction (and related changes elsewhere) could be split into a separate patch.

jrajahalme replied:
Will consider tomorrow.

joestringer commented on pkg/policy/selectorcache.go:
```diff
@@ -533,45 +673,3 @@ func (sc *SelectorCache) RemoveIdentitiesFQDNSelectors(fqdnSels []api.FQDNSelect
-// selector cache. If an identical EndpointSelector has already been
-// cached, the corresponding CachedSelector is returned, otherwise one
-// is created and added to the cache.
-func AddIdentitySelector(user CachedSelectionUser, selector api.EndpointSelector) (cachedSelector CachedSelector, added bool) {
```
Is it possible to shift these bits that remove the singleton to a separate patch?

jrajahalme replied:
Maybe, will consider tomorrow.

joestringer reviewed (dismissed):
I think that this PR would benefit from some consideration for how to break it down, handle a subset of the changes in separate PRs, tidy them up, get them in. Even github/my browser thinks so (while trying to review my feedback for this summary):

![unresponsive_pr](https://user-images.githubusercontent.com/1243336/57656400-184c4180-758d-11e9-93d9-a04cad6e692f.png)

Easy pickings are the one that's already factored out (#7962), the `Equals` -> `DeepEquals` stuff (some of which is part of the big patch in the middle), including the new vendor dependency. I didn't see any other obvious big refactors so far, though I have asked questions below (particularly towards the end) to question whether things can be reworked a bit to make the changes more independent.

The refactoring towards the end looks like an improvement. I'm not sure I agree with the endpoint functions requiring initialization with the `repo` parameter, but we're also still triggering `SelectorPolicy` resolution from the endpoint logic so I think it's reasonable for this to be the case as an intermediate step.

I have a small number of questions about particular parts, like the merging of port rules, and l4filter creation with respect to handling of zero-length selectors. I've tried to limit trivial feedback as the comment feed on this PR is already getting pretty crowded.

ianvernon commented on bpf/Makefile:
```diff
@@ -4,7 +4,7 @@ include ../Makefile.defs
 
 SUBDIRS = sockops
 
-BPF_SIMPLE = bpf_netdev.o bpf_overlay.o bpf_xdp.o bpf_ipsec.o bpf_network.o bpf_alignchecker.o bpf_hostdev_ingress.c
```
Can you factor this out into a separate PR please?

jrajahalme replied:
Did already: #8015

ianvernon commented on pkg/identity/cache/allocator.go:
```diff
@@ -174,7 +174,15 @@ func IdentityAllocationIsLocal(lbls labels.Labels) bool {
 // allocated via the kvstore.
-func AllocateIdentity(ctx context.Context, lbls labels.Labels) (*identity.Identity, bool, error) {
+func AllocateIdentity(owner IdentityAllocatorOwner, ctx context.Context, lbls labels.Labels) (id *identity.Identity, allocated bool, err error) {
+	defer func() {
```
Add a comment indicating what is going on here - this is not obvious.

jrajahalme replied:
Added:
```
	// Notify the owner of the newly added identities so that the
	// cached identities can be updated ASAP, rather than just
	// relying on the kv-store update events.
```

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -1001,9 +994,6 @@ func (e *Endpoint) syncPolicyMap() error {
 			if err != nil {
 				e.getLogger().WithError(err).Errorf("Failed to add PolicyMap key %s %d", policyKeyToPolicyMapKey.String(), entry.ProxyPort)
 				errors = append(errors, err)
-			} else {
```
Why is this being removed? How will be track the realized state of the PolicyMap for the endpoint ?

jrajahalme replied:
After the policy update is complete, the current policy is exactly the one described by the `desiredPolicy`. Instead of modifying the `realizedPolicy` contents, we simply set the `realizedPolicy` to point to `desiredPolicy`.

jrajahalme replied:
This also handles the otherwise needed revert if any errors happen here.

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -675,6 +676,19 @@ func (e *Endpoint) GetModel() *models.Endpoint {
+		l4Policy = e.desiredPolicy.L4Policy
+	}
+	if l4Policy == nil && e.realizedPolicy != nil {
+		l4Policy = e.realizedPolicy.L4Policy
```
Why would we want realizedPolicy to be desiredPolicy? This doesn't make sense to me.

jrajahalme replied:
This was a remnant from my attempt to clear the desired L4Policy pointer when it has been realized. If no new desired policy was created, then it would have been the same as the realized policy. This approach did not work for other reasons.

Thanks for pointing this out, I removed this, as the e.desiredPolicy.L4Policy is now never cleared.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -231,10 +231,10 @@ func (l7 L7DataMap) addRulesForEndpoints(rules api.L7Rules, endpoints []CachedSe
 	selectorCache.RemoveSelectors(l4, l4.CachedSelectors)
-	l4.CachedSelectors = nil
-	l4.L7RulesPerEp = nil
+	//l4.CachedSelectors = nil
```
Why do these no longer need to be set to nil?

jrajahalme replied:
The L4 policy may still be used (temporarily) as a realized or desired policy of an endpoint, until the endpoint completes regeneration. So we must not change the L4 policy at all, but we still need the remove the L4 Policy from the selector cache users.

This is why I renamed this to "Detach()" (from "Delete()").

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -562,10 +562,6 @@ func (e *Endpoint) runPreCompilationSteps(owner Owner, regenContext *regeneratio
 		close(datapathRegenCtxt.ctCleaned)
```
Please provide more context in the commit message - what was needed to be fixed exactly?

jrajahalme replied:
This is mostly a remnant from my nulling the desired L4 policy when Realizing, I'll revert this.

ianvernon commented on pkg/endpoint/bpf.go:
```diff
@@ -673,27 +669,24 @@ func (e *Endpoint) runPreCompilationSteps(owner Owner, regenContext *regeneratio
+		// At this point, traffic is no longer redirected to the proxy for
+		// now-obsolete redirects, since we synced the updated policy map above.
+		// It's now safe to remove the redirects from the proxy's configuration.
+		finalizeFunc, revertFunc = e.removeOldRedirects(owner, desiredRedirects, datapathRegenCtxt.proxyWaitGroup)
```
What is the need to put this inside this if condition vs. outside as before?

jrajahalme replied:
Will revert this, too.

ianvernon commented on pkg/labels/array.go:
```diff
@@ -198,3 +198,16 @@ func (ls LabelArray) String() string {
 	return res
 }
+
+// Same returns true if the label arrays are the same, i.e., have the same labels in the same order.
```
Unit tests?

jrajahalme replied:
Added.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -233,8 +233,6 @@ func (l7 L7DataMap) addRulesForEndpoints(rules api.L7Rules, endpoints []CachedSe
 // the filter is left to be garbage collected!
```
Please provide a more detailed commit message describing the motivation behind this change.

jrajahalme replied:
Lost context for this, sorry.

jrajahalme committed:
checker: Add support for using google/go-cmp

reflect.DeepEquals does not support cyclical data structures. While it
is good to avoid them, sometimes they are useful. Add support for
using google/go-cmp, which allows data types to define their own
Equal() function, which is used for comparisons. When such a function
omits comparing the cyclical internals, cyclical data structures can
be compared.

One example of a cyclical data structure is a cache that tracks the
users of a cached entity. The cached entity itself is the same
regardless of how many users it has, so an Equal() function can omit
comparing the tracked users.

This is used by the forthcoming Selector Cache.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Add special treatment for namespace

Most selectors care about the namespace, so it makes sense to
short-circuit the selection if the namespace does not match. The
benefit of this needs to be benchmarked, though.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Update SelectorCache functionality.

Add new functionality needed for Selector Cache
integration. Streamline locking by avoiding defer in simple functions.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

Commit:
labels: Add Same() for comparing two LabelArrays.

This helps avoid converting label arrays to strings for comparison purposes.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
identity: Initialize well-known identities before the policy repository.

NewPolicyRepository() now grabs the IdentityCache, so all well-known
identities must be initialized before then. This commit removes
InitWellKnownIdentities() from InitIdentityAllocator() so that the
well-known identities can be initialized earlier.

Unfortunately the well-known identities depend on the runtime
configuration and thus can't be statically initialized. Before this
commit there was a potential for race in initializing and using the
well-known identities map. Now that the well-known identities map is
initialized earlier, this race is less likely to be a problem.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
selectorcache: Remove globals.

Global singletons make concurrent unit testing difficult, so remove
the global SelectorCache, and add a SelectorCache pointer to
policy.Repository.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

Commit:
identity: notify owner on identity creation / releasing

KV-store event may take some time to come in. Issue SelectorCache
updates on all locally made allocations as soon as they happen so that
the cache is up-to-date sooner.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

Commit:
endpoint: update Owner interface to include new functions

Add new functions as part of Owner interface:

`GetNodeSuffix` and `UpdateIdentities` are needed to implement the
`IdentityAllocatorOwner` interface for releasing identities / updating
information about identities for the SelectorCache.

Signed-off-by: Ian Vernon <ian@cilium.io>

jrajahalme committed:
policy: Use selector cache in policy computation

api.EndpointSelectors in the computed L4Policy are replaced with
CachedSelectors. L4Filter member 'Endpoints' is renamed to
'CachedSelectors', which now owns the reference to the cached
selector. Duplicates are not inserted into CachedSelectors. *L4Filter
is the owner of the cached selectors. *CachedSelectors are also used
as keys in L7RulesPerEp. Each such key also exists in
'CachedSelectors'.

L4Policies are detached from the selector cache when the policy is
removed from the PolicyCache. From this on the L4Policies may still be
temporarily referenced by the Endpoints as their regeneration
progresses.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Remove more dead code.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Make policy cache a member of Repository, hide internals

The main motivation for this refactor was to make selector policy
remember the selector cache so that clean-up is easier and we don't
need to pass the selector cache pointer through too many function
signarures.

This resulted into a cascade of refactors:

- Endpoint creation functions now need a *Repository parameter, as
  they create EndpointPolicies that now need a *SelectorCache.

- Remove the globalPolicyCache singleton in favor of making Repository
  own a policy cache. This allows concurrent unit testing with
  different Repositories not share a policy cache. This required to
  make policy cache public ('policyCache' -> 'PolicyCache')

- Integrate pkg/policy/distillery to pkg/policy to avoid circular
  import

- Made SelectorPolicy struct private ('selectorPolicy') to avoid name
  collision with the SelectorPolicy interface. This also helps ensure
  policy package internals not leaking elsewhere.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

ianvernon reviewed (approved):
Approving - I think that we will eventually have to revisit the need of making endpoint aware of the repository, and what we want the endpoint structure to contain w.r.t. policy-related fields. This is out of scope of this PR, though.

joestringer reviewed (approved):
At this point, my thoughts on this PR are that it has some potential to be disruptive when merged, but without merging it, running it, bashing on it, we won't get any further. Multiple people have reviewed, the feedback is largely addressed, it's passed CI here and a few times through @ianvernon's PR on top. If @jrajahalme is happy with it, let's get it in so we can continue to iterate on top. This will unblock the upcoming FQDN selector cache improvements, benchmarking to assess progress, and other structural changes that may be necessary.
